File: Backend/StrategyEngine/Engine/RiskManager.fs
module RiskManager

open AST
open EngineTypes
open System

type Scenario = {
    Name: string
    PriceMultiplier: float
    VolMultiplier: float
}

type SellableUnit = 
    | SingleUnit of PositionInstance
    | CompositeUnit of {| GroupId: Guid; Positions: PositionInstance list; BuyDate: int |}

let private stressScenarios = [
    { Name = "Flat"; PriceMultiplier = 1.0; VolMultiplier = 1.0 }
    { Name = "Crash -20%"; PriceMultiplier = 0.8; VolMultiplier = 1.3 }
    { Name = "Crash -50%"; PriceMultiplier = 0.5; VolMultiplier = 2.0 }
    { Name = "Crash -99%"; PriceMultiplier = 0.01; VolMultiplier = 1.0 }
    { Name = "Moon +20%"; PriceMultiplier = 1.2; VolMultiplier = 0.9 }
    { Name = "Moon +100%"; PriceMultiplier = 2.0; VolMultiplier = 1.5 }
]

let private getSellableUnits (portfolio: Portfolio) : SellableUnit list =
    let grouped = 
        portfolio.Positions
        |> List.filter (fun p -> p.GroupId.IsSome)
        |> List.groupBy (fun p -> p.GroupId.Value)
        |> List.map (fun (gid, posList) -> 
            let groupDate = posList |> List.map (fun p -> p.BuyDate) |> List.min
            CompositeUnit {| GroupId = gid; Positions = posList; BuyDate = groupDate |}
        )
        |> List.sortBy (fun c -> match c with CompositeUnit u -> u.BuyDate | _ -> 0)

    let singles = 
        portfolio.Positions
        |> List.filter (fun p -> p.GroupId.IsNone)
        |> List.sortBy (fun p -> p.BuyDate)
        |> List.map SingleUnit

    singles @ grouped

let private removeUnit (portfolio: Portfolio) (unit: SellableUnit) : Portfolio =
    let idsToRemove = 
        match unit with
        | SingleUnit p -> Set.singleton p.Id
        | CompositeUnit c -> c.Positions |> List.map (fun p -> p.Id) |> Set.ofList
    
    { portfolio with 
        Positions = portfolio.Positions |> List.filter (fun p -> not (idsToRemove.Contains p.Id)) 
    }

let private getUnderlyingTicker (instrument: ResolvedInstrument) : string =
    match instrument with
    | ResolvedAsset assetRef -> 
        match assetRef with | SimpleAsset t -> t | LeveragedAsset(t, _) -> t
    | ResolvedOption opt -> 
        match opt.Underlying with | SimpleAsset t -> t | LeveragedAsset(t, _) -> t
    | Compound -> "COMPOUND"

let private calculatePositionValue 
    (position: PositionInstance) 
    (scenario: Scenario) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : float =
    
    match position.Instrument with
    | ResolvedAsset assetRef ->
        let ticker = 
            match assetRef with | SimpleAsset t -> t | LeveragedAsset(t, _) -> t
        
        let basePrice = marketData.[ticker].Price
        
        let scenarioPrice = 
            match assetRef with
            | SimpleAsset _ -> basePrice * scenario.PriceMultiplier
            | LeveragedAsset(_, leverage) -> 
                let pctChange = (scenario.PriceMultiplier - 1.0) * leverage
                basePrice * (1.0 + pctChange) |> max 0.0

        scenarioPrice * position.Quantity

    | ResolvedOption opt ->
        let ticker = 
            match opt.Underlying with | SimpleAsset t -> t | LeveragedAsset(t, _) -> t
            
        let basePrice = marketData.[ticker].Price
        let scenarioUnderlyingPrice = basePrice * scenario.PriceMultiplier
        let dte = opt.ExpiryDay - currentDay
        
        // Note: calculateOptionPrice still uses the multiplier logic, but now takes riskFreeRate
        let optionPrice = PricingModels.calculateOptionPrice opt scenarioUnderlyingPrice scenario.VolMultiplier dte riskFreeRate
        optionPrice * position.Quantity * 100.0

    | Compound -> 0.0

let private calculateBuyingPower 
    (portfolio: Portfolio) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : float =

    let currentScenario = { Name = "Current"; PriceMultiplier = 1.0; VolMultiplier = 1.0 }
    let liquidationValue = 
        portfolio.Positions 
        |> List.sumBy (fun p -> calculatePositionValue p currentScenario marketData currentDay riskFreeRate)
    
    let totalEquity = portfolio.Cash + liquidationValue

    let positionsByUnderlying = 
        portfolio.Positions 
        |> List.groupBy (fun p -> getUnderlyingTicker p.Instrument)

    let totalMarginRequirement = 
        positionsByUnderlying
        |> List.sumBy (fun (ticker, positions) ->
            if ticker = "COMPOUND" then 0.0
            else
                let currentVal = 
                    positions 
                    |> List.sumBy (fun p -> calculatePositionValue p currentScenario marketData currentDay riskFreeRate)

                let worstCasePnL = 
                    stressScenarios
                    |> List.map (fun scenario -> 
                        let scenarioVal = 
                            positions 
                            |> List.sumBy (fun p -> calculatePositionValue p scenario marketData currentDay riskFreeRate)
                        scenarioVal - currentVal
                    )
                    |> List.min

                if worstCasePnL < 0.0 then abs(worstCasePnL) else 0.0
        )

    totalEquity - totalMarginRequirement

let private simulateTrade 
    (portfolio: Portfolio) 
    (trade: PrimitiveTrade) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : Portfolio =
    
    let currentScenario = { Name = "Current"; PriceMultiplier = 1.0; VolMultiplier = 1.0 }

    match trade with
    | PrimitiveRebalance _ -> portfolio
    | PrimitiveBuy p ->
        let dummyPos = { 
            Id = Guid.Empty; GroupId = None; DefinitionName = ""; ComponentName = None; ParentId = None;
            BuyPrice = 0.0; BuyDate = 0; Quantity = p.Quantity; Instrument = p.Instrument 
        }
        let cost = calculatePositionValue dummyPos currentScenario marketData currentDay riskFreeRate
        
        let newPos = {
            Id = Guid.NewGuid()
            GroupId = None 
            DefinitionName = Option.defaultValue "" p.DefinitionName
            ComponentName = p.ComponentName
            ParentId = None
            BuyPrice = (cost / p.Quantity)
            BuyDate = currentDay
            Quantity = p.Quantity
            Instrument = p.Instrument
        }

        { portfolio with 
            Cash = portfolio.Cash - cost
            Positions = newPos :: portfolio.Positions 
        }

    | PrimitiveSell p ->
        let dummyPos = { 
            Id = Guid.Empty; GroupId = None; DefinitionName = ""; ComponentName = None; ParentId = None;
            BuyPrice = 0.0; BuyDate = 0; Quantity = p.Quantity; Instrument = p.Instrument 
        }
        let proceeds = calculatePositionValue dummyPos currentScenario marketData currentDay riskFreeRate

        let rec reducePositions (remainingToSell: float) (positions: PositionInstance list) (acc: PositionInstance list) =
            if remainingToSell <= 0.0 then (acc @ positions)
            else
                match positions with
                | [] -> 
                    let shortPos = {
                        Id = Guid.NewGuid(); GroupId = None; DefinitionName = "SHORT"; ComponentName = None; ParentId = None;
                        BuyPrice = (proceeds / p.Quantity); BuyDate = currentDay; Quantity = -remainingToSell; Instrument = p.Instrument
                    }
                    shortPos :: acc
                | pos :: rest ->
                    if pos.Instrument = p.Instrument then
                        if pos.Quantity > remainingToSell then
                            let reducedP = { pos with Quantity = pos.Quantity - remainingToSell }
                            acc @ (reducedP :: rest)
                        else
                            reducePositions (remainingToSell - pos.Quantity) rest acc
                    else
                        reducePositions remainingToSell rest (pos :: acc)

        let sortedPositions = portfolio.Positions |> List.sortBy (fun p -> p.BuyDate)
        let newPositions = reducePositions p.Quantity sortedPositions []

        { portfolio with 
            Cash = portfolio.Cash + proceeds
            Positions = newPositions 
        }

let validateTrades 
    (trades: PrimitiveTrade list) 
    (portfolio: Portfolio) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : Result<unit, string> =
    
    let hypotheticalPortfolio = 
        trades 
        |> List.fold (fun port trade -> simulateTrade port trade marketData currentDay riskFreeRate) portfolio

    let bp = calculateBuyingPower hypotheticalPortfolio marketData currentDay riskFreeRate

    if bp >= 0.0 then Ok ()
    else Error $"Insufficient Buying Power. Post-trade BP would be ${bp:F2}"

let calculateMaxQuantity 
    (tradesForOneUnit: PrimitiveTrade list) 
    (portfolio: Portfolio) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : int =
    
    let currentBP = calculateBuyingPower portfolio marketData currentDay riskFreeRate

    let oneUnitPortfolio = 
        tradesForOneUnit 
        |> List.fold (fun port trade -> simulateTrade port trade marketData currentDay riskFreeRate) portfolio

    let oneUnitBP = calculateBuyingPower oneUnitPortfolio marketData currentDay riskFreeRate
    let costPerUnit = currentBP - oneUnitBP

    if costPerUnit <= 0.0 then 10000
    else int (Math.Floor(currentBP / costPerUnit))

let analyzeAndPlanRebalance 
    (target: ActionTarget) 
    (targetPercentage: float) 
    (portfolio: Portfolio) 
    (marketData: Map<Identifier, MarketDataPoint>)
    (currentDay: int) 
    (riskFreeRate: float) // <--- ADDED
    : RebalanceAnalysis =

    let currentScenario = { Name = "Current"; PriceMultiplier = 1.0; VolMultiplier = 1.0 }
    
    let currentPortfolioValue = 
        portfolio.Cash + (portfolio.Positions |> List.sumBy (fun p -> calculatePositionValue p currentScenario marketData currentDay riskFreeRate))
    
    let targetValue = currentPortfolioValue * (targetPercentage / 100.0)
    
    let (targetPrice, resolvedTargetInstrument) = 
        match target with
        | AssetTarget assetRef ->
            let ticker = match assetRef with | SimpleAsset t -> t | LeveragedAsset(t, _) -> t
            let price = marketData.[ticker].Price 
            (price, ResolvedAsset assetRef)
        | IdentifierTarget _ -> 
             failwith "Identifier targets not supported for rebalancing logic yet"

    let targetQty = if targetPrice = 0.0 then 0.0 else targetValue / targetPrice
    let cost = targetQty * targetPrice

    let targetPositionInstance = {
        Id = Guid.NewGuid()
        GroupId = None
        DefinitionName = "REBALANCE_TARGET"
        ComponentName = None
        ParentId = None
        BuyPrice = targetPrice
        BuyDate = currentDay
        Quantity = targetQty
        Instrument = resolvedTargetInstrument
    }

    let goalPortfolio = { 
        portfolio with 
            Positions = targetPositionInstance :: portfolio.Positions 
            Cash = portfolio.Cash - cost 
    }

    let startBP = calculateBuyingPower goalPortfolio marketData currentDay riskFreeRate

    if startBP >= 0.0 then
        { IsAchievable = true; PreparatoryTrades = []; DebugReason = "Sufficient Capital" }
    else
        let candidates = 
            getSellableUnits goalPortfolio
            |> List.filter (fun unit -> 
                match unit with
                | SingleUnit p -> p.Id <> targetPositionInstance.Id
                | CompositeUnit c -> c.Positions |> List.forall (fun p -> p.Id <> targetPositionInstance.Id)
            )

        let rankedCandidates = 
            candidates
            |> List.map (fun unit ->
                let testPortfolio = removeUnit goalPortfolio unit
                
                let cashBack = 
                    match unit with
                    | SingleUnit p -> calculatePositionValue p currentScenario marketData currentDay riskFreeRate
                    | CompositeUnit c -> c.Positions |> List.sumBy (fun p -> calculatePositionValue p currentScenario marketData currentDay riskFreeRate)
                
                let testPortfolioWithCash = { testPortfolio with Cash = testPortfolio.Cash + cashBack }
                
                let newBP = calculateBuyingPower testPortfolioWithCash marketData currentDay riskFreeRate
                let impact = newBP - startBP
                
                {| Unit = unit; Impact = impact |}
            )
            |> List.filter (fun x -> x.Impact > 0.0)
            |> List.sortByDescending (fun x -> x.Impact)

        let mutable currentBP = startBP
        let mutable unitsToSell = []
        let mutable solved = false

        for candidate in rankedCandidates do
            if not solved then
                unitsToSell <- candidate.Unit :: unitsToSell
                currentBP <- currentBP + candidate.Impact
                if currentBP >= 0.0 then
                    solved <- true

        if solved then
            let trades = 
                unitsToSell 
                |> List.collect (fun unit ->
                    match unit with
                    | SingleUnit p -> 
                        [ PrimitiveSell { Instrument = p.Instrument; Quantity = p.Quantity; ComponentName = p.ComponentName; DefinitionName = Some p.DefinitionName } ]
                    | CompositeUnit c -> 
                        c.Positions |> List.map (fun p -> 
                            PrimitiveSell { Instrument = p.Instrument; Quantity = p.Quantity; ComponentName = p.ComponentName; DefinitionName = Some p.DefinitionName }
                        )
                )
            
            { IsAchievable = true; PreparatoryTrades = trades; DebugReason = "Rebalance Plan Generated" }
        else
            { IsAchievable = false; PreparatoryTrades = []; DebugReason = "Cannot raise sufficient Buying Power" }

File: Frontend/src/assets/fable_build/Engine/RiskManager.js
import { Union, Record } from "../fable_modules/fable-library-js.4.27.0/Types.js";
import { union_type, anonRecord_type, list_type, class_type, int32_type, record_type, float64_type, string_type } from "../fable_modules/fable-library-js.4.27.0/Reflection.js";
import { PrimitiveTrade, SellParams, RebalanceAnalysis, ResolvedInstrument, PositionInstance, Portfolio, PositionInstance_$reflection } from "./EngineTypes.js";
import { singleton as singleton_1, collect, forAll, sortByDescending, fold, empty, head, tail, isEmpty, cons, sumBy, append, filter, min, map, sortBy, ofArray } from "../fable_modules/fable-library-js.4.27.0/List.js";
import { disposeSafe, getEnumerator, equals, stringHash, comparePrimitives } from "../fable_modules/fable-library-js.4.27.0/Util.js";
import { List_groupBy } from "../fable_modules/fable-library-js.4.27.0/Seq2.js";
import { defaultArg, value } from "../fable_modules/fable-library-js.4.27.0/Option.js";
import { FSharpSet__Contains, singleton, ofList } from "../fable_modules/fable-library-js.4.27.0/Set.js";
import { calculateOptionPrice } from "../Simulation/PricingModels.js";
import { FSharpMap__get_Item } from "../fable_modules/fable-library-js.4.27.0/Map.js";
import { max } from "../fable_modules/fable-library-js.4.27.0/Double.js";
import { newGuid } from "../fable_modules/fable-library-js.4.27.0/Guid.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.4.27.0/Result.js";

export class Scenario extends Record {
    constructor(Name, PriceMultiplier, VolMultiplier) {
        super();
        this.Name = Name;
        this.PriceMultiplier = PriceMultiplier;
        this.VolMultiplier = VolMultiplier;
    }
}

export function Scenario_$reflection() {
    return record_type("RiskManager.Scenario", [], Scenario, () => [["Name", string_type], ["PriceMultiplier", float64_type], ["VolMultiplier", float64_type]]);
}

export class SellableUnit extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["SingleUnit", "CompositeUnit"];
    }
}

export function SellableUnit_$reflection() {
    return union_type("RiskManager.SellableUnit", [], SellableUnit, () => [[["Item", PositionInstance_$reflection()]], [["Item", anonRecord_type(["BuyDate", int32_type], ["GroupId", class_type("System.Guid")], ["Positions", list_type(PositionInstance_$reflection())])]]]);
}

const stressScenarios = ofArray([new Scenario("Flat", 1, 1), new Scenario("Crash -20%", 0.8, 1.3), new Scenario("Crash -50%", 0.5, 2), new Scenario("Crash -99%", 0.01, 1), new Scenario("Moon +20%", 1.2, 0.9), new Scenario("Moon +100%", 2, 1.5)]);

function getSellableUnits(portfolio) {
    const grouped = sortBy((c) => {
        if (c.tag === 1) {
            return c.fields[0].BuyDate | 0;
        }
        else {
            return 0;
        }
    }, map((tupledArg) => {
        const posList = tupledArg[1];
        return new SellableUnit(1, [{
            BuyDate: min(map((p_2) => p_2.BuyDate, posList), {
                Compare: comparePrimitives,
            }),
            GroupId: tupledArg[0],
            Positions: posList,
        }]);
    }, List_groupBy((p_1) => value(p_1.GroupId), filter((p) => (p.GroupId != null), portfolio.Positions), {
        Equals: (x, y) => (x === y),
        GetHashCode: stringHash,
    })), {
        Compare: comparePrimitives,
    });
    return append(map((Item) => (new SellableUnit(0, [Item])), sortBy((p_4) => p_4.BuyDate, filter((p_3) => (p_3.GroupId == null), portfolio.Positions), {
        Compare: comparePrimitives,
    })), grouped);
}

function removeUnit(portfolio, unit) {
    const idsToRemove = (unit.tag === 1) ? ofList(map((p_1) => p_1.Id, unit.fields[0].Positions), {
        Compare: comparePrimitives,
    }) : singleton(unit.fields[0].Id, {
        Compare: comparePrimitives,
    });
    return new Portfolio(portfolio.Cash, filter((p_2) => !FSharpSet__Contains(idsToRemove, p_2.Id), portfolio.Positions), portfolio.CompositeRegistry, portfolio.TaxLots, portfolio.TaxLiabilityYTD, portfolio.RealizedGainsYTD);
}

function getUnderlyingTicker(instrument) {
    switch (instrument.tag) {
        case 1: {
            const matchValue = instrument.fields[0].Underlying;
            if (matchValue.tag === 1) {
                return matchValue.fields[0];
            }
            else {
                return matchValue.fields[0];
            }
        }
        case 2:
            return "COMPOUND";
        default: {
            const assetRef = instrument.fields[0];
            if (assetRef.tag === 1) {
                return assetRef.fields[0];
            }
            else {
                return assetRef.fields[0];
            }
        }
    }
}

function calculatePositionValue(position, scenario, marketData, currentDay, riskFreeRate) {
    let matchValue_1;
    const matchValue = position.Instrument;
    switch (matchValue.tag) {
        case 1: {
            const opt = matchValue.fields[0];
            return (calculateOptionPrice(opt, FSharpMap__get_Item(marketData, (matchValue_1 = opt.Underlying, (matchValue_1.tag === 1) ? matchValue_1.fields[0] : matchValue_1.fields[0])).Price * scenario.PriceMultiplier, scenario.VolMultiplier, opt.ExpiryDay - currentDay, riskFreeRate) * position.Quantity) * 100;
        }
        case 2:
            return 0;
        default: {
            const assetRef = matchValue.fields[0];
            const basePrice = FSharpMap__get_Item(marketData, (assetRef.tag === 1) ? assetRef.fields[0] : assetRef.fields[0]).Price;
            return ((assetRef.tag === 1) ? max(0, basePrice * (1 + ((scenario.PriceMultiplier - 1) * assetRef.fields[1]))) : (basePrice * scenario.PriceMultiplier)) * position.Quantity;
        }
    }
}

function calculateBuyingPower(portfolio, marketData, currentDay, riskFreeRate) {
    const currentScenario = new Scenario("Current", 1, 1);
    return (portfolio.Cash + sumBy((p) => calculatePositionValue(p, currentScenario, marketData, currentDay, riskFreeRate), portfolio.Positions, {
        GetZero: () => 0,
        Add: (x, y) => (x + y),
    })) - sumBy((tupledArg) => {
        const positions = tupledArg[1];
        if (tupledArg[0] === "COMPOUND") {
            return 0;
        }
        else {
            const currentVal = sumBy((p_2) => calculatePositionValue(p_2, currentScenario, marketData, currentDay, riskFreeRate), positions, {
                GetZero: () => 0,
                Add: (x_2, y_2) => (x_2 + y_2),
            });
            const worstCasePnL = min(map((scenario) => (sumBy((p_3) => calculatePositionValue(p_3, scenario, marketData, currentDay, riskFreeRate), positions, {
                GetZero: () => 0,
                Add: (x_3, y_3) => (x_3 + y_3),
            }) - currentVal), stressScenarios), {
                Compare: comparePrimitives,
            });
            if (worstCasePnL < 0) {
                return Math.abs(worstCasePnL);
            }
            else {
                return 0;
            }
        }
    }, List_groupBy((p_1) => getUnderlyingTicker(p_1.Instrument), portfolio.Positions, {
        Equals: (x_1, y_1) => (x_1 === y_1),
        GetHashCode: stringHash,
    }), {
        GetZero: () => 0,
        Add: (x_5, y_5) => (x_5 + y_5),
    });
}

function simulateTrade(portfolio, trade, marketData, currentDay, riskFreeRate) {
    const currentScenario = new Scenario("Current", 1, 1);
    switch (trade.tag) {
        case 0: {
            const p = trade.fields[0];
            const cost = calculatePositionValue(new PositionInstance("00000000-0000-0000-0000-000000000000", undefined, "", undefined, undefined, 0, 0, p.Quantity, p.Instrument), currentScenario, marketData, currentDay, riskFreeRate);
            return new Portfolio(portfolio.Cash - cost, cons(new PositionInstance(newGuid(), undefined, defaultArg(p.DefinitionName, ""), p.ComponentName, undefined, cost / p.Quantity, currentDay, p.Quantity, p.Instrument), portfolio.Positions), portfolio.CompositeRegistry, portfolio.TaxLots, portfolio.TaxLiabilityYTD, portfolio.RealizedGainsYTD);
        }
        case 1: {
            const p_1 = trade.fields[0];
            const proceeds = calculatePositionValue(new PositionInstance("00000000-0000-0000-0000-000000000000", undefined, "", undefined, undefined, 0, 0, p_1.Quantity, p_1.Instrument), currentScenario, marketData, currentDay, riskFreeRate);
            const reducePositions = (remainingToSell_mut, positions_mut, acc_mut) => {
                reducePositions:
                while (true) {
                    const remainingToSell = remainingToSell_mut, positions = positions_mut, acc = acc_mut;
                    if (remainingToSell <= 0) {
                        return append(acc, positions);
                    }
                    else if (!isEmpty(positions)) {
                        const rest = tail(positions);
                        const pos = head(positions);
                        if (equals(pos.Instrument, p_1.Instrument)) {
                            if (pos.Quantity > remainingToSell) {
                                return append(acc, cons(new PositionInstance(pos.Id, pos.GroupId, pos.DefinitionName, pos.ComponentName, pos.ParentId, pos.BuyPrice, pos.BuyDate, pos.Quantity - remainingToSell, pos.Instrument), rest));
                            }
                            else {
                                remainingToSell_mut = (remainingToSell - pos.Quantity);
                                positions_mut = rest;
                                acc_mut = acc;
                                continue reducePositions;
                            }
                        }
                        else {
                            remainingToSell_mut = remainingToSell;
                            positions_mut = rest;
                            acc_mut = cons(pos, acc);
                            continue reducePositions;
                        }
                    }
                    else {
                        return cons(new PositionInstance(newGuid(), undefined, "SHORT", undefined, undefined, proceeds / p_1.Quantity, currentDay, -remainingToSell, p_1.Instrument), acc);
                    }
                    break;
                }
            };
            return new Portfolio(portfolio.Cash + proceeds, reducePositions(p_1.Quantity, sortBy((p_2) => p_2.BuyDate, portfolio.Positions, {
                Compare: comparePrimitives,
            }), empty()), portfolio.CompositeRegistry, portfolio.TaxLots, portfolio.TaxLiabilityYTD, portfolio.RealizedGainsYTD);
        }
        default:
            return portfolio;
    }
}

export function validateTrades(trades, portfolio, marketData, currentDay, riskFreeRate) {
    const bp = calculateBuyingPower(fold((port, trade) => simulateTrade(port, trade, marketData, currentDay, riskFreeRate), portfolio, trades), marketData, currentDay, riskFreeRate);
    if (bp >= 0) {
        return new FSharpResult$2(0, [undefined]);
    }
    else {
        return new FSharpResult$2(1, [`Insufficient Buying Power. Post-trade BP would be $%P(F2)`]);
    }
}

export function calculateMaxQuantity(tradesForOneUnit, portfolio, marketData, currentDay, riskFreeRate) {
    const currentBP = calculateBuyingPower(portfolio, marketData, currentDay, riskFreeRate);
    const costPerUnit = currentBP - calculateBuyingPower(fold((port, trade) => simulateTrade(port, trade, marketData, currentDay, riskFreeRate), portfolio, tradesForOneUnit), marketData, currentDay, riskFreeRate);
    if (costPerUnit <= 0) {
        return 10000;
    }
    else {
        return ~~Math.floor(currentBP / costPerUnit) | 0;
    }
}

export function analyzeAndPlanRebalance(target, targetPercentage, portfolio, marketData, currentDay, riskFreeRate) {
    const currentScenario = new Scenario("Current", 1, 1);
    const targetValue = (portfolio.Cash + sumBy((p) => calculatePositionValue(p, currentScenario, marketData, currentDay, riskFreeRate), portfolio.Positions, {
        GetZero: () => 0,
        Add: (x, y) => (x + y),
    })) * (targetPercentage / 100);
    let patternInput;
    if (target.tag === 1) {
        throw new Error("Identifier targets not supported for rebalancing logic yet");
    }
    else {
        const assetRef = target.fields[0];
        patternInput = [FSharpMap__get_Item(marketData, (assetRef.tag === 1) ? assetRef.fields[0] : assetRef.fields[0]).Price, new ResolvedInstrument(0, [assetRef])];
    }
    const targetPrice = patternInput[0];
    const targetQty = (targetPrice === 0) ? 0 : (targetValue / targetPrice);
    const cost = targetQty * targetPrice;
    const targetPositionInstance = new PositionInstance(newGuid(), undefined, "REBALANCE_TARGET", undefined, undefined, targetPrice, currentDay, targetQty, patternInput[1]);
    const goalPortfolio = new Portfolio(portfolio.Cash - cost, cons(targetPositionInstance, portfolio.Positions), portfolio.CompositeRegistry, portfolio.TaxLots, portfolio.TaxLiabilityYTD, portfolio.RealizedGainsYTD);
    const startBP = calculateBuyingPower(goalPortfolio, marketData, currentDay, riskFreeRate);
    if (startBP >= 0) {
        return new RebalanceAnalysis(true, empty(), "Sufficient Capital");
    }
    else {
        let currentBP = startBP;
        let unitsToSell = empty();
        let solved = false;
        const enumerator = getEnumerator(sortByDescending((x_3) => x_3.Impact, filter((x_2) => (x_2.Impact > 0), map((unit_1) => {
            const testPortfolio = removeUnit(goalPortfolio, unit_1);
            return {
                Impact: calculateBuyingPower(new Portfolio(testPortfolio.Cash + ((unit_1.tag === 1) ? sumBy((p_4) => calculatePositionValue(p_4, currentScenario, marketData, currentDay, riskFreeRate), unit_1.fields[0].Positions, {
                    GetZero: () => 0,
                    Add: (x_1, y_1) => (x_1 + y_1),
                }) : calculatePositionValue(unit_1.fields[0], currentScenario, marketData, currentDay, riskFreeRate)), testPortfolio.Positions, testPortfolio.CompositeRegistry, testPortfolio.TaxLots, testPortfolio.TaxLiabilityYTD, testPortfolio.RealizedGainsYTD), marketData, currentDay, riskFreeRate) - startBP,
                Unit: unit_1,
            };
        }, filter((unit) => {
            if (unit.tag === 1) {
                return forAll((p_2) => (p_2.Id !== targetPositionInstance.Id), unit.fields[0].Positions);
            }
            else {
                return unit.fields[0].Id !== targetPositionInstance.Id;
            }
        }, getSellableUnits(goalPortfolio)))), {
            Compare: comparePrimitives,
        }));
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const candidate = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
                if (!solved) {
                    unitsToSell = cons(candidate.Unit, unitsToSell);
                    currentBP = (currentBP + candidate.Impact);
                    if (currentBP >= 0) {
                        solved = true;
                    }
                }
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        if (solved) {
            return new RebalanceAnalysis(true, collect((unit_2) => {
                if (unit_2.tag === 1) {
                    return map((p_6) => (new PrimitiveTrade(1, [new SellParams(p_6.Instrument, p_6.Quantity, p_6.ComponentName, p_6.DefinitionName)])), unit_2.fields[0].Positions);
                }
                else {
                    const p_5 = unit_2.fields[0];
                    return singleton_1(new PrimitiveTrade(1, [new SellParams(p_5.Instrument, p_5.Quantity, p_5.ComponentName, p_5.DefinitionName)]));
                }
            }, unitsToSell), "Rebalance Plan Generated");
        }
        else {
            return new RebalanceAnalysis(false, empty(), "Cannot raise sufficient Buying Power");
        }
    }
}

